lib/utils.js

```diff
@@ -16,3 +23,27 @@ export const weightedRandom = values => {
 
   return () => samples[Math.floor(Math.random() * samples.length)];
 };
+
+export const isReactComponents = components => {
+  let isValid = true;
+
+  if (components && _.isArray(components) && components.length > 0) {
+    components.forEach(component => {
```

Comment:
Anytime you can exit a loop early (here for example you can stop iterating as soon as you find one invalid "component"), you should use a classic for loop (`for (...;...;...) { ... }`), which you can interrupt early with a `break` or `return`. With forEach, you can't interrupt the loop execution and run through items needlessly. It's probably not a noticeable performance impact here, but it adds up, and in other places it might have an impact, so it's good to have the habit.

Reply:
thanks for the suggestion. should we always use classic loop for array? or can we just use `forEach` loop if there is no need to use `break` statement?

Reply:
I think the classic loop is not that complicated, and it avoids the closure scoping + the function calls, so it's undeniably less computation, and it allows for break/return, so there are not many downsides. That said, with modern JS runtimes like V8, and small collections to loop over, forEach is fine. So just do this, if you know it's in a critical path (during user interaction on browser for example), or that the number of items to loop over is large, or the closure context is large (many local vars that will be captured by the closure), go for classic, otherwise forEach is ok.
Damn, this is answer is longer than you should think about the problem. That's why I do classic 99% of the time, just don't need to think about it. 😆

Reply:
indeed it is a long answer 😃 but thanks for the answers.